flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/serde/LogicalTypeJsonSerializer.java

```diff
@@ -109,349 +123,358 @@ public void serialize(
-            }
+    /**
+     * Checks whether the given type can be serialized as a compact string created from {@link
+     * LogicalType#asSerializableString()}.
+     */
+    private static class CompactSerializationChecker extends LogicalTypeDefaultVisitor<Boolean> {
+
+        private final boolean serializeCatalogObjects;
+
+        CompactSerializationChecker(boolean serializeCatalogObjects) {
+            this.serializeCatalogObjects = serializeCatalogObjects;
         }
 
-        jsonGenerator.writeObject(rawType.asSerializableString());
-    }
+        @Override
+        public Boolean visit(CharType charType) {
+            return charType.getLength() > 0;
+        }
 
-    private void serializeDataTypeForDataView(
-            String key,
-            DataType dataType,
-            JsonGenerator jsonGenerator,
-            SerializerProvider serializerProvider)
-            throws IOException {
-        jsonGenerator.writeFieldName(key);
-        jsonGenerator.writeStartObject();
-        jsonGenerator.writeBooleanField(
-                FIELD_NAME_IS_INTERNAL_TYPE, DataTypeUtils.isInternal(dataType));
-        jsonGenerator.writeFieldName(FIELD_NAME_TYPE_NAME);
-        LogicalType logicalType = LogicalTypeDataTypeConverter.toLogicalType(dataType);
-        serialize(logicalType, jsonGenerator, serializerProvider);
-        jsonGenerator.writeEndObject();
+        @Override
+        public Boolean visit(VarCharType varCharType) {
+            return varCharType.getLength() > 0;
+        }
+
+        @Override
+        public Boolean visit(BinaryType binaryType) {
+            return binaryType.getLength() > 0;
+        }
+
+        @Override
+        public Boolean visit(VarBinaryType varBinaryType) {
+            return varBinaryType.getLength() > 0;
+        }
+
+        @Override
+        public Boolean visit(TimestampType timestampType) {
+            return timestampType.getKind() == TimestampKind.REGULAR;
+        }
+
+        @Override
+        public Boolean visit(ZonedTimestampType zonedTimestampType) {
+            return zonedTimestampType.getKind() == TimestampKind.REGULAR;
+        }
+
+        @Override
+        public Boolean visit(LocalZonedTimestampType localZonedTimestampType) {
+            return localZonedTimestampType.getKind() == TimestampKind.REGULAR;
+        }
+
+        @Override
+        public Boolean visit(DistinctType distinctType) {
+            // catalog-based distinct types are always string serializable,
+            // however, depending on the configuration, we serialize the entire type
+            return !serializeCatalogObjects;
+        }
+
+        @Override
+        public Boolean visit(StructuredType structuredType) {
+            // catalog-based structured types are always string serializable,
+            // however, depending on the configuration, we serialize the entire type
+            return structuredType.getObjectIdentifier().isPresent() && !serializeCatalogObjects;
```

Comment:
In case the structured type doesn't have an object identifier, aren't we able to compat serialize structured types which children are compat serializable? In the same fashion we do for row.

Reply:
Not sure if I understand this question correctly, But this is exactly what we do. The children will have a compact representation. However, there is no string representation defined in the SQL standard like `ROW`. We would need to make up syntax that should actually be defined by a type DDL.